Rework attempt to simplify collectd.conf change.

OrestVoznyy opened the pull request:
Added update_config_file() and few servicing functions.
As result, there are runtime accessible sections of collectd.conf
as lhost.ui.collectd.<section_data> structures; and a single method
recreating collectd.conf file based on that data.

All existing methods were kept for compatibility with many existing
test cases that are using those complex calls.

Updated code to use dict instead of list of tuples.
For consistency, added dependent files:
  - ipmitool.py
  - linux_host_bash.py

Resolved merge conflicts. Updated copyright dates.

Change-Id: I3956764c47b86b18416ec1554e1d268692b06aee
Signed-off-by: Voznyy, OrestX <orestx.voznyy@intel.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/taf3/taf/11)
<!-- Reviewable:end -->

OrestVoznyy committed:
Rework attempt to simplify collectd.conf change.
Added update_config_file() and few servicing functions.
As result, there are runtime accessible sections of collectd.conf
as lhost.ui.collectd.<section_data> structures; and a single method
recreating collectd.conf file based on that data.

All existing methods were kept for compatibility with many existing
test cases that are using those complex calls.

Updated code to use dict instead of list of tuples.
For consistency, added dependent files:
  - ipmitool.py
  - linux_host_bash.py

Resolved merge conflicts. Updated copyright dates.

Change-Id: I3956764c47b86b18416ec1554e1d268692b06aee
Signed-off-by: Voznyy, OrestX <orestx.voznyy@intel.com>

OrestVoznyy committed:
Rework attempt to simplify collectd.conf change.
Added update_config_file() and few servicing functions.
As result, there are runtime accessible sections of collectd.conf
as lhost.ui.collectd.<section_data> structures; and a single method
recreating collectd.conf file based on that data.

All existing methods were kept for compatibility with many existing
test cases that are using those complex calls.

Updated code to use dict instead of list of tuples.
For consistency, added dependent files:
  - ipmitool.py
  - linux_host_bash.py

Resolved merge conflicts. Updated copyright dates.

Reworked data structure and transforming mechanism to provide
more flexibility in plugins configuration definition.
Updated examples of usage, addressing review comments.

Restored collectd init argument cli_set for compatibility.
Removed unchanged ipmitool.py from review.

Updated TAF modules and unittests according to proposed collectd rework.

Refactored logics of text block construction using context managers,
following suggestion in review comments.
Updated negative scenarioi unittest.

Addressed review comment related to service_lib instance method call.
Fixed copyright year ranges.

Removed start/stop/restart methods.
Instead added overloaded call method and method for getting attribute
value from service manager.

Change-Id: I3956764c47b86b18416ec1554e1d268692b06aee
Signed-off-by: Voznyy, OrestX <orestx.voznyy@intel.com>
Signed-off-by: Orest Voznyy <orestx.voznyy@intel.com>

orestkreminskyi commented on taf/testlib/linux/collectd.py:
```diff
@@ -34,160 +34,99 @@
-env.lhost[1].ui.collectd.enable_defaults.csv()
-env.lhost[1].ui.collectd.change_param.csv(DataDir='"/path/to/logs"')
-env.lhost[1].ui.collectd.insert_param.csv(DataDir='"/path/to/logs"')
+env.lhost[1].ui.collectd.update_config_file()
```
Can you please provide complete example for collectd.conf update.
e.g.:
env.lhost[1].ui.global_options = {'Interval': 1}
env.lhost[1].ui.collectd.loadplugin_tags = {'csv': {}}
env.lhost[1].ui.collectd.loadplugin_tags = {'csv': {}}
env.lhost[1].ui.collectd.plugins_options = {'csv': {'DataDir': '"/path/to/some/dir"'}}
env.lhost[1].ui.collectd.update_config_file()

rbbratta commented:
Reivew scrub: needs review

esmacgil commented on taf/testlib/linux/collectd.py:
```diff
@@ -31,163 +31,151 @@
 
 
-def command_generator(action, plugin, collectd_conf):
+def build_tagged_section(config_data, indent_level=-1, res=''):
```
We should use a StringIO buffer and I'd be inclined to do the indentation and tagging with context managers:

```
def build_tagged_section(config_data, indent=None, buffer=None):
    ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for ...:
        if ...:
            with indent, TagContext(buffer, indent, *k.split(' ', 1)):
                buffer.write(build_tagged_section(v, indent, buffer))
        elif ...:
            buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v)
        else:
            buffer.write(PARAM_BOILERPLATE.indent=indent.pad, name=k, value=v)
    return buffer.getvalue()
```

and then somewhere:

```
class IndentationContext(object):
    def __init__(self, char=' ', count=4, buffer=None):
        super().__init__()
        self._char = char
        self._count = count
        self._pad = ''
        self._buffer = buffer
    @property
    def pad(self):
        return self._pad
    def __enter__(self):
        self._pad += self._char * self._count
    def __exit__(self, *args, **kwargs):
        self._pad = self._pad[:-self._count]

class TagContext(object):
    def __init__(self, buffer, indent, *tag_args):
        super().__init__()
        self._buffer = buffer
        self._indent = indent
        assert tag_args
        self._tag_args = tag_args
    def __enter__(self):
        self._buffer.write(self._indent.pad)
        self._buffer.write('<')
        self._buffer.write(' '.join(self._tag_args))
        self._buffer.write('>')
 def __exit__(self, *args, **kwargs):
        self._buffer.write(self._indent.pad)
        self._buffer.write('</')
        self._buffer.write(self._tag_args[0])
        self._buffer.write('>')
```

OrestVoznyy replied:
I've checked suggested change and in progress of applying it.
Though I doubt that `IntentationContext` will serve properly.
The side effect of it seems to be initial indent of 4 whitespaces, while no indentation is expected for top level tag.
Currently I don't see any elegant way out to fix it.
So I wonder if I miss it, or I may merge indentation handling into TagContext logics. 
Please let me know your opinion.

Current context managers I come with, based on your suggestion:
```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        with indent:
            if isinstance(v, collections.Mapping):
                with TagContext(buffer, indent, *k.split(' ', 1)):
                    build_tagged_section(v, indent, buffer)
            elif not isinstance(v, str) and isinstance(v, collections.Iterable):
                buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
            else:
                buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()


class IndentationContext(object):
    def __init__(self, char=' ', count=4, buffer=None):
        super().__init__()
        self._char = char
        self._count = count
        self._pad = ''
        self._buffer = buffer

    @property
    def pad(self):
        return self._pad

    def __enter__(self):
        self._pad += self._char * self._count

    def __exit__(self, *args, **kwargs):
        self._pad = self._pad[:-self._count]


class TagContext(object):
    def __init__(self, buffer, indent, *tag_args):
        super().__init__()
        self._buffer = buffer
        self._indent = indent
        assert tag_args
        self._tag_args = tag_args

    def __enter__(self):
        self._buffer.write(self._indent.pad)
        self._buffer.write('<')
        self._buffer.write(' '.join(self._tag_args))
        self._buffer.write('>\n')

    def __exit__(self, *args, **kwargs):
        self._buffer.write(self._indent.pad)
        self._buffer.write('</')
        self._buffer.write(self._tag_args[0])
        self._buffer.write('>\n')
```
Please note that `__enter__()` and `__exit__()` methods of TagContext were updated with newline character after closing tag bracket to enforce multiline text that was missing in suggested "sample" code.

esmacgil commented on unittests/test_collectd.py:
```diff
@@ -13,79 +13,45 @@
+    def test_collect_no_plugins_config(self):
+        self.collectd_instance = collectd.Collectd(self.cli_send_mock, self.collectd_conf)
+        self.collectd_instance.plugins_config = {}
+        try:
```
Please use `pytest.raises()`.  (http://doc.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions)

```
with pytest.raises(CustomException, 'CustomException not raised...'):
    self.collectd_instance.update_config_file()
```

OrestVoznyy replied:
OK.

esmacgil commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/ipmitool.py, line 3 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFo9IitnYyDI1CJUY:-KceFo9IitnYyDI1CJUZ:b-e6896x) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/ipmitool.py#L3)):*
> ```Python
> #!/usr/bin/env python
> """
> @copyright Copyright (c) 2015 - 2017, Intel Corporation.
> ```

is this done because we forgot to do it in a previous commit of this year, where this file changed?

---

*[taf/testlib/ui_onpss_shell/ui_onpss_shell.py, line 2 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFwBnZlXPmMQtF4ai:-KceFwBnZlXPmMQtF4aj:b-8j2vis) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/ui_onpss_shell/ui_onpss_shell.py#L2)):*
> ```Python
> """
> @copyright Copyright (c) 2015 - 2016, Intel Corporation.
> ```

update year range

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

rbbratta commented:
Review scrub: please address comments

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11#-:-Kd7p427bYWzXybHvgIh:b-exdv18)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Reviewed 2 of 5 files at r1.
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-KeFSjC9chfKQVhxtOh8:b-17h8iz) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, OrestVoznyy wrote…</i></summary><blockquote>

I've checked suggested change and in progress of applying it.
Though I doubt that `IntentationContext` will serve properly.
The side effect of it seems to be initial indent of 4 whitespaces, while no indentation is expected for top level tag.
Currently I don't see any elegant way out to fix it.
So I wonder if I miss it, or I may merge indentation handling into TagContext logics. 
Please let me know your opinion.

Current context managers I come with, based on your suggestion:
```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        with indent:
            if isinstance(v, collections.Mapping):
                with TagContext(buffer, indent, *k.split(' ', 1)):
                    build_tagged_section(v, indent, buffer)
            elif not isinstance(v, str) and isinstance(v, collections.Iterable):
                buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
            else:
                buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()


class IndentationContext(object):
    def __init__(self, char=' ', count=4, buffer=None):
        super().__init__()
        self._char = char
        self._count = count
        self._pad = ''
        self._buffer = buffer

    @property
    def pad(self):
        return self._pad

    def __enter__(self):
        self._pad += self._char * self._count

    def __exit__(self, *args, **kwargs):
        self._pad = self._pad[:-self._count]


class TagContext(object):
    def __init__(self, buffer, indent, *tag_args):
        super().__init__()
        self._buffer = buffer
        self._indent = indent
        assert tag_args
        self._tag_args = tag_args

    def __enter__(self):
        self._buffer.write(self._indent.pad)
        self._buffer.write('<')
        self._buffer.write(' '.join(self._tag_args))
        self._buffer.write('>\n')

    def __exit__(self, *args, **kwargs):
        self._buffer.write(self._indent.pad)
        self._buffer.write('</')
        self._buffer.write(self._tag_args[0])
        self._buffer.write('>\n')
```
Please note that `__enter__()` and `__exit__()` methods of TagContext were updated with newline character after closing tag bracket to enforce multiline text that was missing in suggested "sample" code.
</blockquote></details>

I see that your version is re-entering the `IndexContext` each time through the for loop, where my version only re-entered it when a new `Mapping` is found.

```
    for ...:
        if Mapping:
            with indent, TagContext(buffer, indent, *k.split(' ', 1)):
                ...
        elif Iterable and not str:
            ...
        else:
            ...
```

not:

```
    for ...:
        with indent:
            if Mapping:
                with TagContext(buffer, indent, *k.split(' ', 1)):
                    ...
            elif Iterable and not str:
                ...
            else:
                ...
```

My version is the way it is for two reasons, first is that the `indent_level += 1` and `indent_level -= 1` are only in the `if Mapping` branch.  The second is that it allows us to combine the two contexts into a single indentation block.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Reviewed 2 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[unittests/test_collectd.py, line 52 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KeF_mng6yKlfyXL5k-w:-KeF_mng6yKlfyXL5k-x:bcevpjn) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/unittests/test_collectd.py#L52)):*
> ```Python
>         self.collectd_instance = collectd.Collectd(self.cli_send_mock, self.collectd_conf)
>         self.collectd_instance.plugins_config = {}
>         try:
> ```

Please use `pytest.raises`:

```
with pytest.raises(CustomException, 'CustomException not raised in case...'):
    self.collectd_instance.update_config_file()
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

OrestVoznyy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-KeIccZ44YfNFBbsgS8u:b9lzed4) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

I see that your version is re-entering the `IndexContext` each time through the for loop, where my version only re-entered it when a new `Mapping` is found.

```
    for ...:
        if Mapping:
            with indent, TagContext(buffer, indent, *k.split(' ', 1)):
                ...
        elif Iterable and not str:
            ...
        else:
            ...
```

not:

```
    for ...:
        with indent:
            if Mapping:
                with TagContext(buffer, indent, *k.split(' ', 1)):
                    ...
            elif Iterable and not str:
                ...
            else:
                ...
```

My version is the way it is for two reasons, first is that the `indent_level += 1` and `indent_level -= 1` are only in the `if Mapping` branch.  The second is that it allows us to combine the two contexts into a single indentation block.
</blockquote></details>

Following code (seems to exactly follow your suggestion)
```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            with indent, TagContext(buffer, indent, *k.split(' ', 1)):
                build_tagged_section(v, indent, buffer)
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
        else:
            buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()
```
produces following output (which differs from expected one):
```
Interval 3
AutoLoadPlugins false
    <LoadPlugin cpu>
    </LoadPlugin>
    <LoadPlugin "csv">
    </LoadPlugin>
    <LoadPlugin "python">
    Globals true
    Interval 5
    </LoadPlugin>
    <Plugin "csv">
    DataDir "/tmp/csv_data/"
    </Plugin>
    <Plugin "python">
    ModulePath "/tmp/"
    Interactive false
    Import "python_module_name"
        <Module "python_module_name">
        Test arg1
        </Module>
    </Plugin>
``` 
Expected output is presented in module docstrings.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-KeLHymC4vrWEsfomR4D:boq7nvj) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, OrestVoznyy wrote…</i></summary><blockquote>

Following code (seems to exactly follow your suggestion)
```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            with indent, TagContext(buffer, indent, *k.split(' ', 1)):
                build_tagged_section(v, indent, buffer)
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
        else:
            buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()
```
produces following output (which differs from expected one):
```
Interval 3
AutoLoadPlugins false
    <LoadPlugin cpu>
    </LoadPlugin>
    <LoadPlugin "csv">
    </LoadPlugin>
    <LoadPlugin "python">
    Globals true
    Interval 5
    </LoadPlugin>
    <Plugin "csv">
    DataDir "/tmp/csv_data/"
    </Plugin>
    <Plugin "python">
    ModulePath "/tmp/"
    Interactive false
    Import "python_module_name"
        <Module "python_module_name">
        Test arg1
        </Module>
    </Plugin>
``` 
Expected output is presented in module docstrings.
</blockquote></details>

I think the `-1` default value for `indent_level` was what I failed to account for.  I swapped the order of entry to the two contexts and I'm getting what you show in the docstrings:

```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    else:
        buffer, init_str = StringIO(), str(buffer)
        buffer.write(init_str)
    try:
        indent = IndentationContext(buffer=buffer, count=int(indent))   # the conversion to int should be done by __init__
    except (TypeError, ValueError):
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            with TagContext(buffer, indent, *k.split(' ', 1)), indent:
                build_tagged_section(v, indent, buffer)
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
        else:
            buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()
```

Of course, I'm not sure why there is so much arithmetic in the original version.  If the default value of `indent_level` is `0` and the only `+1` is done to it within the recursion call, then all the other arithmetic is unnecessary:

```
def build_tagged_section_no_inc_dec(config_data, indent_level=0, res=''):
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            tag, space, name = k.partition(' ')
            res += TAGGED_BOILERPLATE.format(indent=INDENT * indent_level,
                                             tag=tag, space=space, name=name,
                                             params=build_tagged_section(v, indent_level + 1))
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            res += ''.join(PARAM_BOILERPLATE.format(indent=INDENT * indent_level, name=k, value=x) for x in v)
        else:
            res += PARAM_BOILERPLATE.format(indent=INDENT * indent_level, name=k, value=v)
    return res
```

I still like the context manager version, due to the maintaining of the indent pad, instead of re-multiplying it each use.  I definitely prefer the string buffer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-KeLOlzVY_vEcIJXNnb-:brngeyq) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

I think the `-1` default value for `indent_level` was what I failed to account for.  I swapped the order of entry to the two contexts and I'm getting what you show in the docstrings:

```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    else:
        buffer, init_str = StringIO(), str(buffer)
        buffer.write(init_str)
    try:
        indent = IndentationContext(buffer=buffer, count=int(indent))   # the conversion to int should be done by __init__
    except (TypeError, ValueError):
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            with TagContext(buffer, indent, *k.split(' ', 1)), indent:
                build_tagged_section(v, indent, buffer)
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
        else:
            buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()
```

Of course, I'm not sure why there is so much arithmetic in the original version.  If the default value of `indent_level` is `0` and the only `+1` is done to it within the recursion call, then all the other arithmetic is unnecessary:

```
def build_tagged_section_no_inc_dec(config_data, indent_level=0, res=''):
    for k, v in config_data.items():
        if isinstance(v, collections.Mapping):
            tag, space, name = k.partition(' ')
            res += TAGGED_BOILERPLATE.format(indent=INDENT * indent_level,
                                             tag=tag, space=space, name=name,
                                             params=build_tagged_section(v, indent_level + 1))
        elif not isinstance(v, str) and isinstance(v, collections.Iterable):
            res += ''.join(PARAM_BOILERPLATE.format(indent=INDENT * indent_level, name=k, value=x) for x in v)
        else:
            res += PARAM_BOILERPLATE.format(indent=INDENT * indent_level, name=k, value=v)
    return res
```

I still like the context manager version, due to the maintaining of the indent pad, instead of re-multiplying it each use.  I definitely prefer the string buffer.
</blockquote></details>

I made a last minute change to the first `if-else` block and forgot.  The `else` statement really needs to be `elif not isinstance(buffer, StringIO)`.  And perhaps the `if` statement should be `if buffer is None`.  Depends on if you want to handle callers passing in _false_ values as the buffer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

orestkreminskyi commented:
Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

orestkreminskyi commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[taf/testlib/linux/collectd.py, line 186 at r2](https://reviewable.io:443/reviews/taf3/taf/11#-KenEFSj_rkdqVZ88mqr:-KenEFSkfMapr3xTDao3:b-psnxbz) ([raw file](https://github.com/taf3/taf/blob/604e48471e042c44bb4cedae0b9a264553ed46b9/taf/testlib/linux/collectd.py#L186)):*
> ```Python
>         self.send_command = cli_send_command
>         self.collectd_conf = collectd_conf if collectd_conf else self.DEFAULT_COLLECTD_CONF
>         self.service_manager = service_lib.specific_service_manager_factory(self.SERVICE, self.send_command)
> ```

Change service manager to the latest version:
`service_lib.SpecificServiceManager(self.SERVICE, self.send_command)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Still some copyright year ranges and a change requested by Orest K.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-KeoFQDte62HoN3svwt8:bnfp4nl) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

I made a last minute change to the first `if-else` block and forgot.  The `else` statement really needs to be `elif not isinstance(buffer, StringIO)`.  And perhaps the `if` statement should be `if buffer is None`.  Depends on if you want to handle callers passing in _false_ values as the buffer.
</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

*[unittests/test_collectd.py, line 52 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KeF_mng6yKlfyXL5k-w:-KeoG9wCnSiPjnx7hIP6:bg7igkv) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/unittests/test_collectd.py#L52)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

Please use `pytest.raises`:

```
with pytest.raises(CustomException, 'CustomException not raised in case...'):
    self.collectd_instance.update_config_file()
```

</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11#-:-KeoGIay2HRcdtl8Kb_f:b-izcg02)*
<!-- Sent from Reviewable.io -->

OrestVoznyy commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 110 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-Kc_DP3tzYPylSKcKzDs-r1-110:-Kes7klFy23HMNj3VUfT:bhfiutb) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/collectd.py#L110)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

:lgtm:
</blockquote></details>

Addressed in patch sets 12.

---

*[taf/testlib/linux/collectd.py, line 186 at r2](https://reviewable.io:443/reviews/taf3/taf/11#-KenEFSj_rkdqVZ88mqr:-Kes7XxkK75XNwsFRACG:b-m16fak) ([raw file](https://github.com/taf3/taf/blob/604e48471e042c44bb4cedae0b9a264553ed46b9/taf/testlib/linux/collectd.py#L186)):*
<details><summary><i>Previously, orestkreminskyi wrote…</i></summary><blockquote>

Change service manager to the latest version:
`service_lib.SpecificServiceManager(self.SERVICE, self.send_command)`
</blockquote></details>

Nice catch, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> Still some copyright range issues.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[taf/testlib/linux/ipmitool.py, line 3 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFo9IitnYyDI1CJUY:-Kf9-81sVdU9YH7pk8oI:bjihnsr) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/ipmitool.py#L3)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

is this done because we forgot to do it in a previous commit of this year, where this file changed?
</blockquote></details>

Yes? No?

---

*[taf/testlib/ui_onpss_shell/ui_onpss_shell.py, line 2 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFwBnZlXPmMQtF4ai:-Kf9-D65mVEJzpSlcBF2:b-fddpwg) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/ui_onpss_shell/ui_onpss_shell.py#L2)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

update year range
</blockquote></details>

please update year range

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11#-:-Kf8zn9YODjHURO1tcdj:b-3kqoxy)*
<!-- Sent from Reviewable.io -->

OrestVoznyy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[taf/testlib/linux/ipmitool.py, line 3 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFo9IitnYyDI1CJUY:-KfAsVhUT5Yuxl1hkpXK:butf1qx) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/linux/ipmitool.py#L3)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

Yes? No?
</blockquote></details>

No, this is a leftover after code roll back. I will remove it.

---

*[taf/testlib/ui_onpss_shell/ui_onpss_shell.py, line 2 at r1](https://reviewable.io:443/reviews/taf3/taf/11#-KceFwBnZlXPmMQtF4ai:-KfAsfS0RZ8-ObMLS9Q7:b2nhktp) ([raw file](https://github.com/taf3/taf/blob/db90498efef8700031025ae5a7143f86e93ececf/taf/testlib/ui_onpss_shell/ui_onpss_shell.py#L2)):*
<details><summary><i>Previously, esmacgil (Edward MacGillivray) wrote…</i></summary><blockquote>

please update year range
</blockquote></details>

OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

esmacgil commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11#-:-KfEMqm8gF0bhroRj_q0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

orestkreminskyi commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[taf/testlib/linux/collectd.py, line 207 at r4](https://reviewable.io:443/reviews/taf3/taf/11#-KfLqL43ciLdY_yWC3iL:-KfLqL43ciLdY_yWC3iM:b-co9sx3) ([raw file](https://github.com/taf3/taf/blob/dd436cbfc9349e09e731fcef68c5fb3167b6a9e4/taf/testlib/linux/collectd.py#L207)):*
> ```Python
>         @brief  Restart collectd service
>         """
>         return self.service_manager.restart()
> ```


Can you please refactor service_manager related methods in the same way as it was done in pull request 23?
https://github.com/taf3/taf/pull/23
As a result I would expect something similar as follows instead of `start stop restart` methods:
```
+    def __getattr__(self, name):
 +        """
 +        @brief  Method for getting attribute from service_manager
 +        @param  name:  attribute name
 +        @type  name:  string
 +        """
 +        return getattr(self.service_manager, name)
 +
 +    def __call__(self, cmd, expected_rc):
 +        """
 +        @brief  Overloaded call method
 +        @param  cmd:  command to execute
 +        @type  cmd:  string
 +        @param  expected_rc:  expected return code
 +        @type  expected_rc:  int | set | list | frozenset
 +        @rtype:  named tuple
 +        """
 +        return self.cli_send_command(cmd, expected_rcs=expected_rc)
```
This solution will decrease code and increase collectd capabilities(all methods that are present in service_lib will be accessible in collectd instance.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

OrestVoznyy commented:
Reviewed 2 of 4 files at r2, 1 of 1 files at r3, 1 of 2 files at r4.
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[taf/testlib/linux/collectd.py, line 207 at r4](https://reviewable.io:443/reviews/taf3/taf/11#-KfLqL43ciLdY_yWC3iL:-KfvhZqZN4E8HC010l42:b-896fix) ([raw file](https://github.com/taf3/taf/blob/dd436cbfc9349e09e731fcef68c5fb3167b6a9e4/taf/testlib/linux/collectd.py#L207)):*
<details><summary><i>Previously, orestkreminskyi wrote…</i></summary><blockquote>


Can you please refactor service_manager related methods in the same way as it was done in pull request 23?
https://github.com/taf3/taf/pull/23
As a result I would expect something similar as follows instead of `start stop restart` methods:
```
+    def __getattr__(self, name):
 +        """
 +        @brief  Method for getting attribute from service_manager
 +        @param  name:  attribute name
 +        @type  name:  string
 +        """
 +        return getattr(self.service_manager, name)
 +
 +    def __call__(self, cmd, expected_rc):
 +        """
 +        @brief  Overloaded call method
 +        @param  cmd:  command to execute
 +        @type  cmd:  string
 +        @param  expected_rc:  expected return code
 +        @type  expected_rc:  int | set | list | frozenset
 +        @rtype:  named tuple
 +        """
 +        return self.cli_send_command(cmd, expected_rcs=expected_rc)
```
This solution will decrease code and increase collectd capabilities(all methods that are present in service_lib will be accessible in collectd instance.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11)*
<!-- Sent from Reviewable.io -->

orestkreminskyi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/taf3/taf/11#-:-KfvtsjS9b-jbKJpwtQH:bnfp4nl)*
<!-- Sent from Reviewable.io -->
